Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -86,240 +86,329 @@ local macro "pderiv_simp" : tactic =>
-      - (C (2 * W.a₂) * X 0 ^ 2 * X 2 + C (4 * W.a₄) * X 0 * X 2 ^ 3 + C (6 * W.a₆) * X 2 ^ 5) := by
+lemma polynomialZ_eq : V.polynomialZ =
+    C V.a₁ * X 0 * X 1 + C (3 * V.a₃) * X 1 * X 2 ^ 2
+      - (C (2 * V.a₂) * X 0 ^ 2 * X 2 + C (4 * V.a₄) * X 0 * X 2 ^ 3 + C (6 * V.a₆) * X 2 ^ 5) := by
   rw [polynomialZ, polynomial]
   pderiv_simp
   ring1
 
-lemma eval_polynomialZ (P : Fin 3 → R) : eval P W.polynomialZ =
-    W.a₁ * P x * P y + 3 * W.a₃ * P y * P z ^ 2
-      - (2 * W.a₂ * P x ^ 2 * P z + 4 * W.a₄ * P x * P z ^ 3 + 6 * W.a₆ * P z ^ 5) := by
+lemma eval_polynomialZ (P : Fin 3 → R) : eval P V.polynomialZ =
+    V.a₁ * P x * P y + 3 * V.a₃ * P y * P z ^ 2
+      - (2 * V.a₂ * P x ^ 2 * P z + 4 * V.a₄ * P x * P z ^ 3 + 6 * V.a₆ * P z ^ 5) := by
   rw [polynomialZ_eq]
   eval_simp
 
+variable (V) in
 /-- The proposition that a point representative $(x, y, z)$ in `W` is nonsingular.
 In other words, either $W_X(x, y, z) \ne 0$, $W_Y(x, y, z) \ne 0$, or $W_Z(x, y, z) \ne 0$. -/
 def Nonsingular (P : Fin 3 → R) : Prop :=
-  W.Equation P ∧ (eval P W.polynomialX ≠ 0 ∨ eval P W.polynomialY ≠ 0 ∨ eval P W.polynomialZ ≠ 0)
-
-lemma nonsingular_iff (P : Fin 3 → R) : W.Nonsingular P ↔ W.Equation P ∧
-    (W.a₁ * P y * P z ≠ 3 * P x ^ 2 + 2 * W.a₂ * P x * P z ^ 2 + W.a₄ * P z ^ 4 ∨
-      P y ≠ -P y - W.a₁ * P x * P z - W.a₃ * P z ^ 3 ∨
-      W.a₁ * P x * P y + 3 * W.a₃ * P y * P z ^ 2
-        ≠ 2 * W.a₂ * P x ^ 2 * P z + 4 * W.a₄ * P x * P z ^ 3 + 6 * W.a₆ * P z ^ 5) := by
-  rw [Nonsingular, eval_polynomialX, eval_polynomialY, eval_polynomialZ, sub_ne_zero, sub_ne_zero,
-    ← sub_ne_zero (a := P y)]
-  congr! 4
-  ring1
-
-lemma nonsingular_zero [Nontrivial R] : W.Nonsingular ![1, 1, 0] :=
-  (W.nonsingular_iff ![1, 1, 0]).mpr ⟨W.equation_zero,
-    by simp; by_contra! h; exact one_ne_zero <| by linear_combination -h.1 - h.2.1⟩
-
-lemma nonsingular_zero' [NoZeroDivisors R] {Y : R} (hy : Y ≠ 0) :
-    W.Nonsingular ![Y ^ 2, Y ^ 3, 0] :=
-  (W.nonsingular_iff ![Y ^ 2, Y ^ 3, 0]).mpr ⟨W.equation_zero' Y,
-    by simp [hy]; by_contra! h; exact pow_ne_zero 3 hy <| by linear_combination Y ^ 3 * h.1 - h.2.1⟩
-
-lemma nonsingular_some (X Y : R) : W.Nonsingular ![X, Y, 1] ↔ W.toAffine.Nonsingular X Y := by
-  rw [nonsingular_iff]
-  matrix_simp
-  simp only [W.toAffine.nonsingular_iff, equation_some, and_congr_right_iff,
-    W.toAffine.equation_iff, ← not_and_or, not_iff_not, one_pow, mul_one, Iff.comm, iff_self_and]
-  intro h hX hY
-  linear_combination (norm := ring1) 6 * h - 2 * X * hX - 3 * Y * hY
-
-lemma nonsingular_smul_iff (P : Fin 3 → R) (u : Rˣ) : W.Nonsingular (u • P) ↔ W.Nonsingular P :=
-  have (u : Rˣ) {P : Fin 3 → R} (h : W.Nonsingular <| u • P) : W.Nonsingular P := by
-    rcases (W.nonsingular_iff _).mp h with ⟨h, h'⟩
-    refine (W.nonsingular_iff P).mpr ⟨(W.equation_smul_iff P u).mp h, ?_⟩
+  V.Equation P ∧ (eval P V.polynomialX ≠ 0 ∨ eval P V.polynomialY ≠ 0 ∨ eval P V.polynomialZ ≠ 0)
```

Comment:
I think `IsNonsingular` is a better name for this but again this might not be for this PR. It's probably also the wrong definition for a general ring: one should instead probably ask that the evaluations are units rather than nonzero. Is this an issue?